### PR TITLE
Backport ursim_driver to foxy branch

### LINKED ISSUE #270: ursim_driver missing on the foxy branch
Does it need to be back ported, perhaps?

### PR DESCRIPTION
**Pull request by destogl**
resolves #270

### CODE REVIEW

**Comment by urrsk** [comment]
In stead of building the ursim docker, we could use the image on docker hub.
See:
- https://hub.docker.com/repository/docker/universalrobots/ursim_e-series
- https://hub.docker.com/repository/docker/universalrobots/ursim_cb3

**Review by livanov93 — APPROVED** [review]
It looks good to me. I am not sure about @urrsk comment, should we use simplify the ursim dockerizing by pulling it from dockerhub.
I think it is fine like this now. And when we do the simplification, we can do another backport.
@destogl @urrsk Do you agree?

File: ur_robot_driver/resources/ursim_driver/driver/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,86 @@
+# -- BEGIN LICENSE BLOCK ----------------------------------------------
```
Comment (urrsk) [inline ur_robot_driver/resources/ursim_driver/driver/Dockerfile:1]: Please change it to BSD3

File: ur_robot_driver/resources/ursim_driver/ursim/aux/ros_control.urscript [hunk 2]
```diff
@@ -0,0 +1,143 @@
+{{BEGIN_REPLACE}}
```
Comment (urrsk) [inline ur_robot_driver/resources/ursim_driver/ursim/aux/ros_control.urscript:1]: Why is this file needed. It should use the one from the client_library!

File: ur_robot_driver/resources/ursim_driver/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+   ports:
+      - "5900:5900"
+      - "29999:29999"
+      - "8080:8080"
+      - "30001-30004:30001-30004"
+   volumes:
+      - ./ursim/aux/.vol/:/ursim/programs
+   build:
+    context: ursim/.
+    dockerfile: g5/Dockerfile
```
Comment (fmauch) [inline ur_robot_driver/resources/ursim_driver/docker-compose.yml:41]: We could use the upstream docker images released by UR by now. But that's not a must.

File: ur_robot_driver/resources/ursim_driver/ursim/aux/.vol/urcap_ros_control.script [hunk 4]
```diff
@@ -0,0 +1,180 @@
+def urcap_ros_control():
```
Comment (fmauch) [inline ur_robot_driver/resources/ursim_driver/ursim/aux/.vol/urcap_ros_control.script:1]: Why is this moved to the robot already? This should come from our driver, right?

File: ur_robot_driver/resources/ursim_driver/ursim/aux/ros_control.urscript (hunk 2 above)
Comment (fmauch) [reply ur_robot_driver/resources/ursim_driver/ursim/aux/ros_control.urscript:1]: agreed

File: ur_robot_driver/resources/ursim_driver/ursim/g3/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,112 @@
+# MIT License
```
Comment (fmauch) [inline ur_robot_driver/resources/ursim_driver/ursim/g3/Dockerfile:1]: I think, using the released docker images would be better.

**Review by fmauch — COMMENTED** [review]
Over all, I think, it would be better to use the released upstream images. In a tutorial package I recently started, I created a rather slim version of an example bringing up preconfigured robots with the urcap installed and a program loaded: https://github.com/UniversalRobots/Universal_Robots_ROS_Tutorials/blob/main/ur_example_dual_robot/resources/docker-compose.yml

However, as this is already the state that we have on the main branch, I would also be OK with merging it like this.

**Comment by fmauch** [comment]
I'll remove this from the Foxy release milestone, as I am still not sure whether we want this in like that. I'd like to change the tests a bit more using the official URSim Dockerfile and I am not sure how much sense it makes to have a docker-compose file in this repo actually building and running the driver...

**Comment by fmauch** [comment]
This will not be done anymore for Foxy.
